### PR TITLE
add liftsim baseline

### PR DESCRIPTION
**Pull request by banma12956**

### CODE REVIEW

File: examples/LiftSim_demo/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# LiftSim基线
+
+## 简介
+
+基于PARL库实现Deep Q-network算法，应用于[RLSchool][rlschool]库中的电梯调度模拟环境[LiftSim][liftsim]。
+
+## 依赖库
+
+- paddlepaddle >= 1.2.1
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/README.md]: PARL requires that paddle is at least 1.5.1 now.

File: examples/LiftSim_demo/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+- parl >= 1.1.1
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/README.md]: parl >= 1.1.2

File: examples/LiftSim_demo/demo.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+A running demo of elevators
+
+Authors: wangfan04(wangfan04@baidu.com)
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/demo.py]: Please remove this information about the author and date.

File: examples/LiftSim_demo/rl_benchmark/agent.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        self._obs_dim = obs_dim
+        self._update_target_steps = 1000
+
+        self._global_step = 0
+        self.exploration_ratio = 0.9
+        self.exploration_decre = 1e-7
+        self.exploration_min = 0.1
+        super(ElevatorAgent, self).__init__(algorithm)
+
+        use_cuda = True if self.gpu_id >= 0 else False
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/rl_benchmark/agent.py]: The attribute `gpu_id` of `parl.Agent` is deprecated, pls use our new API:  https://github.com/PaddlePaddle/PARL/blob/b29a1ec13df4f69c1b758db84c71bb81925ea9c8/examples/GA3C/atari_agent.py#L51

File: examples/LiftSim_demo/rl_benchmark/agent.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            else:
+                action = np.argmax(q_values[i])
+            ret_actions.append(int(action))
+        return ret_actions
+        
+    def predict(self, obs):
+        pred_Q = self.fluid_executor.run(
+            self.pred_program,
+            feed={'obs': obs.astype('float32')},
+            fetch_list=[self._value])  # [0]
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/rl_benchmark/agent.py]: remove `# [0]`.

File: examples/LiftSim_demo/rl_benchmark/dispatcher.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+        self._algorithm = DQN(self._model, hyperparas)
+        self._agent = ElevatorAgent(
+            self._algorithm, self._obs_dim, self._act_dim)
+        self._warm_up_size = 2000
+        self._statistic_freq = 1000
+        self._loss_queue = deque()
+
+    def run_episode(self):
+        # self._agent.restore('./model.ckpt')
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/rl_benchmark/dispatcher.py]: pls remove this line.

File: examples/LiftSim_demo/rl_benchmark/model.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+class RLDispatcherModel(Model):
+    def __init__(self, act_dim):
+        self._act_dim = act_dim
+        self._fc_1 = layers.fc(size=512, act='relu')
+        self._fc_2 = layers.fc(size=256, act='relu')
+        self._fc_3 = layers.fc(size=128, act='tanh')
+        self._output = layers.fc(size=act_dim)
+
+    def value(self, obs):
+        self._h_1 = self._fc_1(obs)
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/rl_benchmark/model.py]: It's unnecessary that  variables like `_h_1`, `_h_2` are attached to the class, please remove `self.`

File: examples/LiftSim_demo/rl_benchmark/model.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import os
+import paddle.fluid as fluid
+from parl import layers
+import numpy as np
+from parl import Model
+from parl import Agent
+from parl.utils import get_gpu_count
+
+
+class RLDispatcherModel(Model):
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/rl_benchmark/model.py]: use `parl.Model` here

File: examples/LiftSim_demo/rl_benchmark/model.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import paddle.fluid as fluid
+from parl import layers
+import numpy as np
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/rl_benchmark/model.py:18]: remove this line

File: examples/LiftSim_demo/rl_benchmark/model.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+from parl import Model
+from parl import Agent
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/rl_benchmark/model.py]: remove the following two lines.

File: examples/LiftSim_demo/wrapper.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# wrapper part modified from
+# https://github.com/openai/gym/blob/master/gym/core.py
+
+from rlschool import LiftSim
+from wrapper_utils import obs_dim, act_dim, mansion_state_preprocessing
+from wrapper_utils import action_idx_to_action
+
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/wrapper.py:21]: All of the wrapper should be a subclass of `gym.Wrapper`. Please refer to what we did [before](http://gitlab.baidu.com/nlp-ol/R2_NIPS2018_ProstheticsChallenge_DDPG/blob/master/env_wrapper.py).

File: examples/LiftSim_demo/wrapper_utils.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+from rlschool import EPSILON, HUGE
+from rlschool import MansionConfig
+from rlschool import MansionManager
+
+
+def discretize(value, n_dim, min_val, max_val):
+    """
+    discretize a value into a vector of n_dim dimension 1-hot representation
+    with the value below min_val being [1, 0, 0, ..., 0]
+    and the value above max_val being [0, 0, ..., 0, 1]
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/wrapper_utils.py:29]: Add `Args` and `Returns` describing the the input and output of the function.

File: examples/LiftSim_demo/wrapper_utils.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+    active_pos = int((value - min_val) / delta + 0.5)
+    active_pos = min(n_dim - 1, active_pos)
+    active_pos = max(0, active_pos)
+    ret_array = [0 for i in range(n_dim)]
+    ret_array[active_pos] = 1.0
+    return ret_array
+
+
+def linear_discretize(value, n_dim, min_val, max_val):
+    """
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_demo/wrapper_utils.py:51]: Add `Args` and `Returns` describing the the input and output of the function.

File: examples/LiftSim_demo/wrapper.py (hunk 11 above)
Comment (banma12956) [reply examples/LiftSim_demo/wrapper.py:21]: The link raises 404 error

Comment (TomorrowIsAnOtherDay) [reply examples/LiftSim_demo/wrapper.py:21]: image
Oh, I forgot that you don't have the permission to view this page.

**Review by TomorrowIsAnOtherDay — APPROVED** [review]

**Pull request merged**